tpcc/check.go

```diff
@@ -3,129 +3,274 @@ package main
+	var err error
+	if rowsLeft, err = db.Query("SELECT w_ytd FROM warehouse ORDER BY w_id"); err != nil {
+		return err
+	}
+
+	if rowsRight, err = db.Query("SELECT SUM(d_ytd) FROM district GROUP BY d_w_id ORDER BY d_w_id"); err != nil {
+		return err
+	}
+
+	var left, right float64
+	var i int
+	for ; rowsLeft.Next() && rowsRight.Next(); i++ {
+		rowsLeft.Scan(&left)
+		rowsRight.Scan(&right)
+
+		if left != right {
+			return errors.Errorf("warehouse %d: warehouse.w_ytd %f != sum(district.d_ytd) %f",
+				i, left, right)
 		}
 	}
+
+	if rowsLeft.Next() || rowsRight.Next() {
+		return errors.Errorf("length mismatch between warehouse.w_ytd and sum(district.d_ytd)")
+	}
+	if err := rowsLeft.Close(); err != nil {
+		return err
+	}
+	if err := rowsRight.Close(); err != nil {
+		return err
+	}
+
+	if i == 0 {
+		return errors.Errorf("zero rows")
+	}
+
+	fmt.Printf("3.3.2.1: %d rows, passed in %f seconds.\n", i, time.Since(start).Seconds())
 	return nil
 }
 
-func checkConsistency(db *sql.DB) error {
-	// Reused checks
+func check3322(db *sql.DB) error {
+	start := time.Now()
 
-	maxNewOIDs := make([]int, *warehouses*10)
-
-	sliceIdx := func(wID, dID int) int {
-		return wID*10 + dID - 1
+	districtRows, err := db.Query("SELECT d_next_o_id FROM district ORDER BY d_w_id, d_id")
+	if err != nil {
+		return err
 	}
-
-	// 3.3.2.1
-	// W_YTD = sum(D_YTD) for all warehouses and districts
-	var wYTD, sumDYTD float64
-	for i := 0; i < *warehouses; i++ {
-		if err := db.QueryRow("SELECT w_ytd FROM warehouse WHERE w_id=$1", i).Scan(&wYTD); err != nil {
-			return err
-		}
-		if err := db.QueryRow("SELECT SUM(d_ytd) FROM district WHERE d_w_id=$1", i).Scan(&sumDYTD); err != nil {
-			return err
-		}
-		if wYTD != sumDYTD {
-			fmt.Printf("check failed: w_ytd=%f != sum(d_ytd)=%f for warehouse %d\n", wYTD, sumDYTD, i)
-		}
+	newOrderRows, err := db.Query("SELECT MAX(no_o_id) FROM tpcc.new_order GROUP BY no_d_id, no_w_id ORDER BY no_w_id, no_d_id")
+	if err != nil {
+		return err
+	}
+	orderRows, err := db.Query("SELECT MAX(o_id) FROM tpcc.order GROUP BY o_d_id, o_w_id ORDER BY o_w_id, o_d_id")
+	if err != nil {
+		return err
 	}
 
-	// 3.3.2.2
-	// D_NEXT_O_ID - 1 = max(O_ID) = max(NO_O_ID) for all districts
-	var dNextOID, maxOID, maxNewOID int
-	if err := forEachWarehouseAndDistrict(func(wID, dID int) error {
-		if err := db.QueryRow(`SELECT d_next_o_id FROM district WHERE d_w_id=$1 AND d_id=$2`,
-			wID, dID,
-		).Scan(&dNextOID); err != nil {
-			return err
-		}
-		if err := db.QueryRow(`SELECT max(o_id) FROM "order" WHERE o_w_id=$1 AND o_d_id=$2`,
-			wID, dID,
-		).Scan(&maxOID); err != nil {
-			return err
-		}
-		if err := db.QueryRow(`SELECT max(no_o_id) FROM new_order WHERE no_w_id=$1 AND no_d_id=$2`,
-			wID, dID,
-		).Scan(&maxNewOID); err != nil {
-			return err
-		}
-		maxNewOIDs[sliceIdx(wID, dID)] = maxNewOID
+	var district, newOrder, order float64
+	var i int
+	for ; districtRows.Next() && newOrderRows.Next() && orderRows.Next(); i++ {
```

Comment:
A 3-way client side merge join? You must have a reason for doing this vs a more complex query.